File: src/lib/rateLimit.js
// Minimal in-memory sliding-window rate limiter for AI-spending routes.
// Per-instance only (resets on deploy/restart) — a stopgap against
// anonymous cost-abuse, not a substitute for real quotas.
const buckets = new Map();

export function rateLimit(req, { limit = 10, windowMs = 60_000 } = {}) {
  const ip =
    req.headers.get("x-forwarded-for")?.split(",")[0]?.trim() ||
    req.headers.get("x-real-ip") ||
    "unknown";

  const now = Date.now();
  const hits = (buckets.get(ip) || []).filter((t) => now - t < windowMs);

  if (hits.length >= limit) {
    buckets.set(ip, hits);
    return Response.json(
      { error: "Too many requests. Try again shortly." },
      { status: 429 }
    );
  }

  hits.push(now);
  buckets.set(ip, hits);

  if (buckets.size > 10_000) buckets.clear();

  return null;
}

File: backend/rate_limit.py
"""Minimal in-memory sliding-window rate limiter.

Per-process only (resets on restart) -- a stopgap against anonymous
cost abuse of AI-spending endpoints, not a substitute for real quotas.
"""

import time

_buckets: dict[str, list[float]] = {}


def is_rate_limited(client_ip: str, limit: int = 10, window_seconds: int = 60) -> bool:
    now = time.monotonic()
    hits = [t for t in _buckets.get(client_ip, []) if now - t < window_seconds]

    if len(hits) >= limit:
        _buckets[client_ip] = hits
        return True

    hits.append(now)
    _buckets[client_ip] = hits

    if len(_buckets) > 10_000:
        _buckets.clear()

    return False
